# Named Entity Recognition using Language Models
This lab will focus on fine-tuning Transformers models for Named Entity Recognition. We will use the Datasets library to quickly load and preprocess the WikiNEuRal dataset, and to get them ready for training with PyTorch.

#### By: Nazia Nafis
---
---

# 1. Using DistilBERT

## Installation
Let's first install the required libraries

In [1]:
!pip install datasets transformers seqeval huggingface_hub

     |████████████████████████████████| 325 kB 5.3 MB/s 
     |████████████████████████████████| 4.0 MB 46.3 MB/s 
     |████████████████████████████████| 43 kB 2.0 MB/s 
     |████████████████████████████████| 77 kB 7.2 MB/s 
     |████████████████████████████████| 1.1 MB 36.1 MB/s 
     |████████████████████████████████| 136 kB 48.8 MB/s 
     |████████████████████████████████| 212 kB 31.0 MB/s 
     |████████████████████████████████| 127 kB 46.5 MB/s 
     |████████████████████████████████| 6.5 MB 32.1 MB/s 
     |████████████████████████████████| 895 kB 36.7 MB/s 
     |████████████████████████████████| 596 kB 38.0 MB/s 
     |████████████████████████████████| 94 kB 3.2 MB/s 
     |████████████████████████████████| 144 kB 40.5 MB/s 
     |████████████████████████████████| 271 kB 49.0 MB/s 
  Created wheel for seqeval: filename=seqeval-1.2.2-py3-none-any.whl size=16180 sha256=0571c92e3b910a168ebfd5534b9b67e0f418cc46ce82f0c793e6242231a13634
  Stored in directory: /root/.cache/pip/whe

In [24]:
!curl -s https://packagecloud.io/install/repositories/github/git-lfs/script.deb.sh | sudo bash
!sudo apt-get install git-lfs
!git lfs install

Detected operating system as Ubuntu/bionic.
Checking for curl...
Detected curl...
Checking for gpg...
Detected gpg...
Running apt-get update... done.
Installing apt-transport-https... done.
Installing /etc/apt/sources.list.d/github_git-lfs.list...done.
Importing packagecloud gpg key... done.
Running apt-get update... done.

The repository is setup! You can now install packages.
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following NEW packages will be installed:
  git-lfs
0 upgraded, 1 newly installed, 0 to remove and 97 not upgraded.
Need to get 6,800 kB of archives.
After this operation, 15.3 MB of additional disk space will be used.
Get:1 https://packagecloud.io/github/git-lfs/ubuntu bionic/main amd64 git-lfs amd64 3.1.2 [6,800 kB]
Fetched 6,800 kB in 1s (12.4 MB/s)
debconf: unable to initialize frontend: Dialog
debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl

## Dataset(s)

We will use the Datasets library to download the data and get the metric we need to use for evaluation (to compare our model to the benchmark). This can be done with the functions load_dataset and load_metric.

The notebook should work with any token classification dataset provided by the Datasets library.

In [2]:
from datasets import load_dataset, load_metric, concatenate_datasets
datasets = load_dataset("Babelscape/wikineural")

Using custom data configuration Babelscape--wikineural-2c05bc228ce4015c


Extracting data files:   0%|          | 0/27 [00:00<?, ?it/s]

Dataset parquet downloaded and prepared to /root/.cache/huggingface/datasets/parquet/Babelscape--wikineural-2c05bc228ce4015c/0.0.0/0b6d5799bb726b24ad7fc7be720c170d8e497f575d02d47537de9a5bac074901. Subsequent calls will reuse this data.


  0%|          | 0/27 [00:00<?, ?it/s]

The WikiNEuRal dataset has train, validation and test splits for each of the 9 languages.

To access an actual element, we need to select a split first, then give an index:

In [3]:
datasets["train_en"][0]

{'lang': 'en',
 'ner_tags': [0, 0, 0, 0, 0, 5, 6, 0, 0, 0, 0, 5, 0, 0],
 'tokens': ['This',
  'division',
  'also',
  'contains',
  'the',
  'Ventana',
  'Wilderness',
  ',',
  'home',
  'to',
  'the',
  'California',
  'condor',
  '.']}

The labels are already coded as integer ids to be easily usable by our model, but the correspondence with the actual categories is stored the following:

In [4]:
label_list = ['O', 'B-PER', 'I-PER', 'B-ORG', 'I-ORG', 'B-LOC', 'I-LOC', 'B-MISC', 'I-MISC']

labels_vocab = {'O': 0, 'B-PER': 1, 'I-PER': 2, 'B-ORG': 3, 'I-ORG': 4, 'B-LOC': 5, 'I-LOC': 6, 'B-MISC': 7, 'I-MISC': 8}
labels_vocab_reverse = {v:k for k,v in labels_vocab.items()}

If we want to train a monolingual model (NER on a single language), we have to use a single language for each split.

In [5]:
train_dataset = datasets["train_en"]
val_dataset = datasets["val_en"]
test_dataset = datasets["test_en"]

This is how we can concatenate multiple dataset splits to train a Multilingual NER model.

In [6]:
train_dataset = concatenate_datasets([datasets["train_de"], 
                                       datasets["train_en"], 
                                       datasets["train_es"]])

val_dataset = concatenate_datasets([datasets["val_de"], 
                                     datasets["val_en"], 
                                     datasets["val_es"]])

test_dataset = concatenate_datasets([datasets["test_de"], 
                                     datasets["test_en"], 
                                    datasets["test_es"]])

## Pre-processing

Before we can feed those texts to our model, we need to preprocess them. This is done by a Transformers Tokenizer which will (as the name indicates) tokenize the inputs (including converting the tokens to their corresponding IDs in the pretrained vocabulary) and put it in a format the model expects, as well as generate the other inputs that model requires.

To do all of this, we instantiate our tokenizer with the AutoTokenizer.from_pretrained method, which will ensure:

 - we get a tokenizer that corresponds to the model architecture we want to use,
 - we download the vocabulary used when pretraining this specific checkpoint.
 
That vocabulary will be cached, so it's not downloaded again the next time we run the cell.

In [7]:
from transformers import AutoTokenizer
    
tokenizer = AutoTokenizer.from_pretrained("distilbert-base-cased") ## We use the BERT tokenizer

Downloading:   0%|          | 0.00/29.0 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/411 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/208k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/426k [00:00<?, ?B/s]

The following assertion ensures that our tokenizer is a fast tokenizers (backed by Rust) from the Tokenizers library. Those fast tokenizers are available for almost all models, and we will need some of the special features they have for our preprocessing.

In [8]:
import transformers
assert isinstance(tokenizer, transformers.PreTrainedTokenizerFast)

We can directly call this tokenizer on one sentence, or we can use the tokenize method to tokenize multiple sentences at once.

In [9]:
tokenizer("Hello, this is one sentence!")

{'input_ids': [101, 8667, 117, 1142, 1110, 1141, 5650, 106, 102], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1]}

This is a good overal tutorial for preprocessing for Transformer. Check this out to learn more about the preprocessing steps.
https://huggingface.co/transformers/preprocessing.html

Here we set the labels of all special tokens to -100 (the index that is ignored by PyTorch) and the labels of all other tokens to the label of the word they come from. Another strategy is to set the label only on the first token obtained from a given word, and give a label of -100 to the other subtokens from the same word. Just change the value of the following flag:

In [10]:
label_all_tokens = False

We're now ready to write the function that will preprocess our samples. We feed them to the tokenizer with the argument truncation=True (to truncate texts that are bigger than the maximum size allowed by the model) and is_split_into_words=True (as seen above). Then we align the labels with the token ids using the strategy we picked:

In [11]:
def tokenize_and_align_labels(examples):
    tokenized_inputs = tokenizer(examples["tokens"], truncation=True, is_split_into_words=True)

    labels = []
    for i, label in enumerate(examples["ner_tags"]):
        word_ids = tokenized_inputs.word_ids(batch_index=i)
        previous_word_idx = None
        label_ids = []
        for word_idx in word_ids:
            # Special tokens have a word id that is None. We set the label to -100 so they are automatically
            # ignored in the loss function.
            if word_idx is None:
                label_ids.append(-100)
            # We set the label for the first token of each word.
            elif word_idx != previous_word_idx:
                label_ids.append(label[word_idx])
            # For the other tokens in a word, we set the label to either the current label or -100, depending on
            # the label_all_tokens flag.
            else:
                label_ids.append(label[word_idx] if label_all_tokens else -100)
            previous_word_idx = word_idx

        labels.append(label_ids)

    tokenized_inputs["labels"] = labels
    return tokenized_inputs

To apply this function on all the sentences (or pairs of sentences) in our dataset, we just use the map method of our dataset object we created earlier. This will apply the function on all the elements of all the splits in dataset, so our training, validation and testing data will be preprocessed in one single command.

In [12]:
train_tokenized = train_dataset.map(tokenize_and_align_labels, batched=True)
val_tokenized = val_dataset.map(tokenize_and_align_labels, batched=True)
test_tokenized = test_dataset.map(tokenize_and_align_labels, batched=True)

  0%|          | 0/268 [00:00<?, ?ba/s]

  0%|          | 0/34 [00:00<?, ?ba/s]

  0%|          | 0/34 [00:00<?, ?ba/s]

Even better, the results are automatically cached by the Datasets library to avoid spending time on this step the next time you run your notebook. The Datasets library is normally smart enough to detect when the function you pass to map has changed (and thus requires to not use the cache data). For instance, it will properly detect if you change the task in the first cell and rerun the notebook. Datasets warns you when it uses cached files, you can pass load_from_cache_file=False in the call to map to not use the cached files and force the preprocessing to be applied again.

Note that we passed batched=True to encode the texts by batches together. This is to leverage the full benefit of the fast tokenizer we loaded earlier, which will use multi-threading to treat the texts in a batch concurrently.

## Fine-tuning

In [13]:
from transformers import AutoModelForTokenClassification, TrainingArguments, Trainer

model = AutoModelForTokenClassification.from_pretrained("distilbert-base-cased", num_labels=len(label_list), label2id=labels_vocab, id2label=labels_vocab_reverse)

Downloading:   0%|          | 0.00/251M [00:00<?, ?B/s]

Some weights of the model checkpoint at distilbert-base-cased were not used when initializing DistilBertForTokenClassification: ['vocab_layer_norm.weight', 'vocab_projector.weight', 'vocab_transform.weight', 'vocab_transform.bias', 'vocab_projector.bias', 'vocab_layer_norm.bias']
- This IS expected if you are initializing DistilBertForTokenClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing DistilBertForTokenClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of DistilBertForTokenClassification were not initialized from the model checkpoint at distilbert-base-cased and are newly initialized: ['classifier.weight', 'classifier.bias']
You should probably TRAIN this 

The warning is telling us we are throwing away some weights (the vocab_transform and vocab_layer_norm layers) and randomly initializing some other (the pre_classifier and classifier layers). This is absolutely normal in this case, because we are removing the head used to pretrain the model on a masked language modeling objective and replacing it with a new head for which we don't have pretrained weights, so the library warns us we should fine-tune this model before using it for inference, which is exactly what we are going to do.

To instantiate a Trainer, we will need to define three more things. The most important is the TrainingArguments, which is a class that contains all the attributes to customize the training. It requires one folder name, which will be used to save the checkpoints of the model, and all other arguments are optional:

In [14]:
model_name = "distilbert-base-multilingual-cased" ## You Change this to any generic language model from HuggingFace which supports fine-tuning.
args = TrainingArguments(                   ## This is where all your hyperparameters are, reduce *_batch_size if you are working with a GPU which has limited memory.
    "wikineural-multilingual-ner",
    evaluation_strategy = "steps",
    learning_rate=2e-5,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    num_train_epochs=1,                     ## may not be enough for a large data, increase this to 5 later in your experiments.
    weight_decay=0.01,
    push_to_hub=True,
    eval_steps=10000,
    save_steps=10000,
)

Here we set the evaluation to be done at the end of each epoch, tweak the learning rate,customize the batch_size and the number of epochs for training, as well as the weight decay.

Then we will need a data collator that will batch our processed examples together while applying padding to make them all the same size (each pad will be padded to the length of its longest example). There is a data collator for this task in the Transformers library, that not only pads the inputs, but also the labels:

In [15]:
from transformers import DataCollatorForTokenClassification

data_collator = DataCollatorForTokenClassification(tokenizer)

The last thing to define for our Trainer is how to compute the metrics from the predictions. Here we will load the seqeval metric (which is commonly used to evaluate results on the CONLL dataset) via the Datasets library.

In [16]:
metric = load_metric("seqeval")

So we will need to do a bit of post-processing on our predictions:

 - select the predicted index (with the maximum logit) for each token
 - convert it to its string label
 - ignore everywhere we set a label of -100
 
The following function does all this post-processing on the result of Trainer.evaluate (which is a namedtuple containing predictions and labels) before applying the metric:

In [17]:
import numpy as np

def compute_metrics(p):
    predictions, labels = p
    predictions = np.argmax(predictions, axis=2)

    # Remove ignored index (special tokens)
    true_predictions = [
        [label_list[p] for (p, l) in zip(prediction, label) if l != -100]
        for prediction, label in zip(predictions, labels)
    ]
    true_labels = [
        [label_list[l] for (p, l) in zip(prediction, label) if l != -100]
        for prediction, label in zip(predictions, labels)
    ]

    results = metric.compute(predictions=true_predictions, references=true_labels)
    return {
        "precision": results["overall_precision"],
        "recall": results["overall_recall"],
        "f1": results["overall_f1"],
        "accuracy": results["overall_accuracy"],
    }

Before we go any further, we need to execute to following steps:
1. Login to the HuggigFace CLI on our system using command ``huggingface-cli login``
2. Get our access token from your user account on HuggingFace at https://huggingface.co/settings/tokens (preferably a write access token)
3. Use this token with the command above to ensure being logged in to HuggingFace.

IF not working on own machine with a GPU and have opened this notebook in Colab / Studio Lab, we can use the also cell below to login. We could have used this cell instead of commandline (cli) login, but where is the fun in that?

In [18]:
from huggingface_hub import notebook_login
notebook_login()

Login successful
Your token has been saved to /root/.huggingface/token
Authenticated through git-credential store but this isn't the helper defined on your machine.
You might have to re-authenticate when pushing to the Hugging Face Hub. Run the following command in your terminal in case you want to set this credential helper as the default

git config --global credential.helper store


Note that we drop the precision/recall/f1 computed for each category and only focus on the overall precision/recall/f1/accuracy.

Then we just need to pass all of this along with our datasets to the Trainer:

In [25]:
trainer = Trainer(
    model,
    args,
    train_dataset=train_tokenized,
    eval_dataset=test_tokenized,
    data_collator=data_collator,
    tokenizer=tokenizer,
    compute_metrics=compute_metrics
)

Cloning https://huggingface.co/nn007/wikineural-multilingual-ner into local empty directory.


Download file pytorch_model.bin:   0%|          | 8.38k/249M [00:00<?, ?B/s]

Download file runs/Apr11_08-14-10_84318bc0e9b7/events.out.tfevents.1649665043.84318bc0e9b7.34.0: 100%|########…

Download file runs/Apr11_12-49-15_8b6cf10e184c/events.out.tfevents.1649681814.8b6cf10e184c.34.0: 100%|########…

Download file runs/Apr11_12-49-15_8b6cf10e184c/1649681814.6651518/events.out.tfevents.1649681814.8b6cf10e184c.…

Download file training_args.bin: 100%|##########| 2.98k/2.98k [00:00<?, ?B/s]

Download file runs/Apr11_08-14-10_84318bc0e9b7/1649665043.6096761/events.out.tfevents.1649665043.84318bc0e9b7.…

Clean file runs/Apr11_08-14-10_84318bc0e9b7/events.out.tfevents.1649665043.84318bc0e9b7.34.0:  14%|#4        |…

Clean file runs/Apr11_12-49-15_8b6cf10e184c/1649681814.6651518/events.out.tfevents.1649681814.8b6cf10e184c.34.…

Clean file runs/Apr11_12-49-15_8b6cf10e184c/events.out.tfevents.1649681814.8b6cf10e184c.34.0:  14%|#4        |…

Clean file training_args.bin:  34%|###3      | 1.00k/2.98k [00:00<?, ?B/s]

Clean file runs/Apr11_08-14-10_84318bc0e9b7/1649665043.6096761/events.out.tfevents.1649665043.84318bc0e9b7.34.…

Clean file pytorch_model.bin:   0%|          | 1.00k/249M [00:00<?, ?B/s]

In [26]:
import os
os.environ["WANDB_DISABLED"] = "true"

In [27]:
! pip install torchvision 
import gc
import torch

gc.collect()
torch.cuda.empty_cache()

In [28]:
trainer.train()

The following columns in the training set  don't have a corresponding argument in `DistilBertForTokenClassification.forward` and have been ignored: ner_tags, lang, tokens. If ner_tags, lang, tokens are not expected by `DistilBertForTokenClassification.forward`,  you can safely ignore this message.
/usr/local/lib/python3.7/dist-packages/transformers/optimization.py:309: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  FutureWarning,
***** Running training *****
  Num examples = 267680
  Num Epochs = 1
  Instantaneous batch size per device = 16
  Total train batch size (w. parallel, distributed & accumulation) = 16
  Gradient Accumulation steps = 1
  Total optimization steps = 16730


The following columns in the evaluation set  don't have a corresponding argument in `DistilBertForTokenClassification.forward` and have been ignored: ner_tags, lang, tokens. If ner_tags, lang, tokens are not expected by `DistilBertForTokenClassification.forward`,  you can safely ignore this message.
***** Running Evaluation *****
  Num examples = 33587
  Batch size = 16
Saving model checkpoint to wikineural-multilingual-ner/checkpoint-10000
Configuration saved in wikineural-multilingual-ner/checkpoint-10000/config.json
Model weights saved in wikineural-multilingual-ner/checkpoint-10000/pytorch_model.bin
tokenizer config file saved in wikineural-multilingual-ner/checkpoint-10000/tokenizer_config.json
Special tokens file saved in wikineural-multilingual-ner/checkpoint-10000/special_tokens_map.json
tokenizer config file saved in wikineural-multilingual-ner/tokenizer_config.json
Special tokens file saved in wikineural-multilingual-ner/special_tokens_map.json


Training completed. Do not fo

TrainOutput(global_step=16730, training_loss=0.045595527507556244, metrics={'train_runtime': 4842.0958, 'train_samples_per_second': 55.282, 'train_steps_per_second': 3.455, 'total_flos': 6036777957669984.0, 'train_loss': 0.045595527507556244, 'epoch': 1.0})

## Evaluation

The evaluate method allows you to evaluate again on the evaluation dataset or on another dataset:

In [29]:
trainer.evaluate()

The following columns in the evaluation set  don't have a corresponding argument in `DistilBertForTokenClassification.forward` and have been ignored: ner_tags, lang, tokens. If ner_tags, lang, tokens are not expected by `DistilBertForTokenClassification.forward`,  you can safely ignore this message.
***** Running Evaluation *****
  Num examples = 33587
  Batch size = 16


{'epoch': 1.0,
 'eval_accuracy': 0.991745638368278,
 'eval_f1': 0.9357391279626243,
 'eval_loss': 0.026128584519028664,
 'eval_precision': 0.9314580031695721,
 'eval_recall': 0.9400597878862887,
 'eval_runtime': 203.5696,
 'eval_samples_per_second': 164.99,
 'eval_steps_per_second': 10.316}

To get the precision/recall/f1 computed for each category now that we have finished training, we can apply the same function as before on the result of the predict method:

In [30]:
predictions, labels, _ = trainer.predict(test_tokenized)
predictions = np.argmax(predictions, axis=2)

# Remove ignored index (special tokens)
true_predictions = [
    [label_list[p] for (p, l) in zip(prediction, label) if l != -100]
    for prediction, label in zip(predictions, labels)
]
true_labels = [
    [label_list[l] for (p, l) in zip(prediction, label) if l != -100]
    for prediction, label in zip(predictions, labels)
]

results = metric.compute(predictions=true_predictions, references=true_labels)
results

The following columns in the test set  don't have a corresponding argument in `DistilBertForTokenClassification.forward` and have been ignored: ner_tags, lang, tokens. If ner_tags, lang, tokens are not expected by `DistilBertForTokenClassification.forward`,  you can safely ignore this message.
***** Running Prediction *****
  Num examples = 33587
  Batch size = 16


{'LOC': {'f1': 0.9466953438353775,
  'number': 17112,
  'precision': 0.9417138892101307,
  'recall': 0.9517297802711547},
 'MISC': {'f1': 0.8708260520474188,
  'number': 10516,
  'precision': 0.8650652153514122,
  'recall': 0.8766641308482312},
 'ORG': {'f1': 0.9330509350139478,
  'number': 9639,
  'precision': 0.9292108241588641,
  'recall': 0.9369229173150742},
 'PER': {'f1': 0.9700516576211339,
  'number': 15252,
  'precision': 0.9674579366114516,
  'recall': 0.9726593233674272},
 'overall_accuracy': 0.991745638368278,
 'overall_f1': 0.9357391279626243,
 'overall_precision': 0.9314580031695721,
 'overall_recall': 0.9400597878862887}

---
---
In this notebook, we saw the performance of **DistilBERT Base Multilingual Cased** model.

We got the following metrics overall:

- Accuracy: 0.99
- F1: 0.94
- Precision: 0.93
- Recall: 0.94